### ISSUE TITLE
Open source loop vectorization infrastructure

### ISSUE THREAD

**Issue by gergo-** [oracle/graal#5439]
**TL;DR**

Make automatic loop vectorization available for just-in-time (JIT) compilations in GraalVM Community Edition by open sourcing relevant parts of the loop vectorizer from GraalVM Enterprise Edition.

**Goals**

Open source compiler phases and supporting infrastructure to enable automatic loop vectorization in GraalVM CE JIT compilations. Users will be able to write natural Java loops and obtain efficient SIMD code for supported loop shapes. This will include many loops that are not vectorized by HotSpot's C2 compiler.

See the blog post on [Enhanced Automated Vectorization](https://medium.com/graalvm/enhanced-automated-vectorization-in-graalvm-76cd99925b6d) for some examples of automatic vectorization in GraalVM.

**Non-Goals**

- It is not a goal to open source GraalVM EE's linear code vectorizer.
- In contrast to support for JIT compilations, it is not a goal to open source all the supporting infrastructure to enable loop vectorization in ahead-of-time (AOT) compilations.
